kythe/extractors/openjdk11/java-wrapper.go

```diff
@@ -0,0 +1,230 @@
+		if root, err := filepath.Abs(base); err == nil {
+			if _, err := os.Stat(root); err == nil {
+				os.Setenv("RUNFILES_DIR", root)
+				if _, err := os.Stat(filepath.Join(root, "MANIFEST")); err == nil {
+					os.Setenv("RUNFILES_MANIFEST_FILE", filepath.Join(root, "MANIFEST"))
+				}
+				return nil
+			}
+		}
+	}
+	return errors.New("unable to setup runfiles")
+}
+
+func findJavaCommand() (string, error) {
+	if val := os.Getenv(javaCommandVar); len(val) > 0 {
+		return val, checkPath(val)
+	}
+	if val := os.Getenv(javaHomeVar); len(val) > 0 {
+		return filepath.Join(val, "bin", "java"), checkPath(val, "bin", "java")
+	}
+	return exec.LookPath("java")
+}
+
+func findExtractorJar() (string, error) {
+	if val := os.Getenv(extractorJarVar); len(val) > 0 {
+		return val, checkPath(val)
+	}
+	return bazel.Runfile(extractorJarPath)
+}
+
+func extractorArgs(args []string, jar string) []string {
+	target := targetJdk()
+	isJavac := false
+	isBootstrap := true
+	var result []string
+	for len(args) > 0 {
+		switch a := shift(&args); a {
+		case "-m", "--module":
+			v := shift(&args)
+			if !strings.HasSuffix(v, ".javac.Main") {
+				// This is not a javac invocation, don't extract.
+				return nil
+			}
+			isJavac = true
+			result = append(result,
+				"--add-exports=jdk.compiler.interim/com.sun.tools.javac.main=ALL-UNNAMED",
+				"--add-exports=jdk.compiler.interim/com.sun.tools.javac.util=ALL-UNNAMED",
+				"--add-exports=jdk.compiler.interim/com.sun.tools.javac.file=ALL-UNNAMED",
+				"--add-exports=jdk.compiler.interim/com.sun.tools.javac.api=ALL-UNNAMED",
+				"--add-exports=jdk.compiler.interim/com.sun.tools.javac.code=ALL-UNNAMED",
+				"-jar", jar, "-Xprefer:source")
+		case "-target":
+			v := shift(&args)
+			if v != target {
+				// This is definitely a bootstrap compilation, don't extract it.
+				return nil
+			}
+			isBootstrap = false
+			result = append(result, a, v)
+		case "--add-modules", "--limit-modules":
+			result = append(result, a, shift(&args)+",java.logging,java.sql")
+		case "--doclint-format":
+			shift(&args)
+		case "-Werror":
+		default:
+			switch {
+			case strings.HasPrefix(a, "-Xplugin:depend"), strings.HasPrefix(a, "-Xlint:"), strings.HasPrefix(a, "-Xdoclint"):
+			case strings.HasPrefix(a, "-Xmx"):
+				result = append(result, "-Xmx3G")
+			default:
+				result = append(result, a)
+			}
+		}
+	}
+	if !isJavac || isBootstrap {
+		return nil
+	}
+	return result
+}
+
+func extractorEnv() []string {
+	env := os.Environ()
+	env = setenvDefault(env, kytheRootVar, ".")
+	env = setenvDefault(env, kytheOutputVar, outputDir())
+	env = setenvDefault(env, kytheTargetVar, moduleName())
+	env = setenvDefault(env, kytheCorpusVar, "openjdk"+targetJdk())
+	return env
+}
+
+func setenvDefault(env []string, key, def string) []string {
+	if val := os.Getenv(key); len(val) == 0 {
+		env = append(env, key+"="+def)
+	}
+	return env
+}
+
+func shift(args *[]string) string {
```

Comment:
An idiomatic form of this should be `func shift(args []string) (string, []string)`

Reply:
Agreed, but that does not play well Go's shadowing and scoping for the intended use.  So either I do this unidiomatic thing or have to remember *never* to use `:=` for the return value (or it's an infinite loop due to shadowing).

Reply:
Idioticized.

Reply:
er, idiomaticized.